language-plutus-core/src/Language/PlutusCore/Constant/Embed.hs

```diff
@@ -0,0 +1,59 @@
+{-# LANGUAGE DataKinds        #-}
+{-# LANGUAGE GADTs            #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Language.PlutusCore.Constant.Embed
+    ( embedStaticBuiltinNameInTerm
+    , embedTypedBuiltinNameInTerm
+    , embedDynamicBuiltinNameInTerm
+    )
+where
+
+import           Language.PlutusCore.Constant.Function
+import           Language.PlutusCore.Constant.Typed
+import           Language.PlutusCore.Core
+import           Language.PlutusCore.MkPlc
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Quote
+
+import           Data.Text                             (append, pack)
+
+{- | Since built-in names don't make sense on their own, we produce a
+   term which wraps the name in a sequence of abstractions and lambdas
+   and then applies the name to the relevant types and variables. This
+   isn't ideal, but it does what's required for testing.
+-}
+embedBuiltinNameInTerm :: TypeScheme uni args res -> BuiltinName -> Term TyName Name uni ()
+embedBuiltinNameInTerm scheme name =
+    let mkVarDecl :: Type TyName uni () -> Quote (VarDecl TyName Name uni ())
+        mkVarDecl ty = do
+          u <- freshUnique
+          let argname = Name (append (pack "arg") (pack . show . unUnique $ u)) u
+          -- This makes the unique explicit in the variable name.  If
+          -- you don't do this then `plc example` outputs examples
+          -- where builtins are applied to multiple arguments with the
+          -- same name.
+          pure $ VarDecl () argname ty
+        mkTyVarDecl :: TyName -> TyVarDecl TyName ()
+        mkTyVarDecl tyname = TyVarDecl () tyname (Type ())
+    in case typeComponentsOfTypeScheme scheme of
+      Nothing -> Prelude.error $ "Malformed type scheme for " ++ show name ++ " in embedBuiltinNameInTerm"
+      -- FIXME: we should really have a proper error here, but that involves running a *lot* of stuff in a monad
+      Just (TypeComponents tynames types _) ->
+          runQuote $ do
+            varDecls <- mapM mkVarDecl types  -- For each argument type, generate a VarDecl for a variable of that type
+            let termArgs = map (Var () . varDeclName) varDecls
+                tyArgs = map (TyVar ()) tynames
+                tyVarDecls = map mkTyVarDecl tynames
+            pure $ mkIterTyAbs tyVarDecls (mkIterLamAbs varDecls (ApplyBuiltin () name tyArgs termArgs))
+
+embedStaticBuiltinNameInTerm :: TypeScheme uni args res -> StaticBuiltinName -> Term TyName Name uni ()
```

Comment:
OK, I did that.  Using `withTypedBuiltinName` meant that I had to add ` (GShow uni, GEq uni, DefaultUni <: uni)` constraints to `embedStaticBuiltinNameInTerm` and a number of things in `Denotation.hs` though. In the end it didn't matter because those constraints always held where the functions were ultimately being called.